tube: fix slow take on busy utubes

DerekBum opened the pull request:
If some of the utubes for tasks at the start of the queue were busy most of the time, `take` would slow down for every other task. This problem is fixed by creating a new space `space_ready`. It contains first task with `READY` status from each utube.

This solution shows great results for the stated problem, but with the cost of slowing the `put` method. Thus, this workaround is disabled by default. To enable it, user should set the `storage_mode = "ready_buffer"` as an option while creating the tube (or `queue.driver.utube.STORAGE_MODE_READY_BUFFER` for `utube` and `queue.driver.utubettl.STORAGE_MODE_READY_BUFFER` for `utubettl`). As example:
```lua
local test_queue = queue.create_tube('test_queue', 'utube',
        {temporary = true, storage_mode = queue.driver.utube.STORAGE_MODE_READY_BUFFER})
```
Or
```lua
local test_queue = queue.create_tube('test_queue', 'utubettl',
        {temporary = true, storage_mode = queue.driver.utubettl.STORAGE_MODE_READY_BUFFER})
```

Also added two benchmarks to compare `storage_mode = "default"` and `storage_mode = "ready_buffer"` (`storage_mode = "default"` is default and disables the workaround).

1. Benchmark for simple `put` and `take` methods. 30k utubes are created with single task each. Task creation time is calculated. After that 30k consumers are calling `take` + `ack`, each in the separate fiber. Time to ack all tasks is calculated. The results are as follows (for comparison I also used `fifo` tube):

|         | put (30k) | take+ack |
|---------|-----------|----------|
| default | 180ms     | 1.6s     |
| ready   | 270ms     | 1.7s     |

As we can see, new utube implementation has slower `put` method.
2. Benchmark for the stated problem. 10 tubes are created. Each contains 1000 task. After that 10 consumers are created (each works on his tube only, one tube -- one consumer). Each consumer will `take`, then `yield` and then `ack` every task from their utube (1000 tasks each).
After that we can also run this benchmark with 10k tasks on each utube, 100k tasks and 150k tasks. But all that with 10 utubes and 10 consumers. The results are as follows:

|         | 1k    | 10k  | 50k  | 150k  |
|---------|-------|------|------|-------|
| default | 53s   | 1.5h | 100h | 1000h |
| ready   | 450ms | 4.7s | 26s  | 72s   |

We can see great time performance improvement.

Same change was also made for `utubettl`. Here is benchmarks results for `utubettl` (on the same benchmarks):

|         | put (30k) | take+ack |
|---------|-----------|----------|
| default | 200ms     | 1.7s     |
| ready   | 320ms     | 1.8s     |

|         | 1k    | 10k  | 50k  | 140k  |
|---------|-------|------|------|-------|
| default | 80s   | 1.6h | 100h | 1000h |
| ready   | 520ms | 5.4s | 28s  | 83s   |

(note that for `busy tasks` were used 140k and not 150k tasks).

Closes #228

DerekBum commented:
Right now only `utube` is fixed. `utubettl` will follow soon.

oleg-jukovec commented on queue/abstract/driver/utube.lua:
```diff
@@ -98,12 +123,61 @@ function method.put(self, data, opts)
 end
 
+local function take_ready(self)
+    for s, task_ready in self.space_ready:pairs({}, { iterator = 'GE' }) do
```
```suggestion
    for s, task_ready in self.space_ready:pairs({}, { iterator = 'GE' }) do
```
It is a little confusing here: we don't need to iterate by the index, by try to peek up a first item and repeat if no success.

DerekBum replied:
But what if the first element is in `READY` state but the utube it is in has `TAKEN` state? Then we will be waiting for utube to leave the `TAKEN` state first, instead of trying the element from the next utube.

oleg-jukovec replied:
We need just to repeat to try to take a first task from the space.

The task in the `_ready` becomes only if the `utube` have not a tasks in the `TAKEN` state.
But we could see a task in the `_ready` space if the `utube` have tasks in the `TAKEN` state due to concurrent updated. But we expect that it gone soon. Maybe we need a `yield` between tries.

oleg-jukovec reviewed (changes requested):
Please, add update the `ready` space for all updates of the `READY` state for a task cases: `release`, as example.

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -53,17 +53,35 @@ function tube.create_space(space_name, opts)
+    if space.v2 and not space_ready then
+        -- Create a space for first ready tasks from each utube.
+        space_ready = box.schema.create_space(space_ready_name, space_opts)
+        space_ready:create_index('task_id', {
+            type = 'tree',
+            parts = {1, num_type()}
+        })
+        space_ready:create_index('utube', {
+            type = 'tree',
+            parts = {2, str_type()}
+        })
```
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/create_index/#lua-data.index_opts.unique
 > Default: true

oleg-jukovec replied:
Let's better declare this clear for the second one and add the configuration flag, since it is not entirely obvious this was done on some purpose.

DerekBum commented:
Added fix for `utubettl` in a different commit.

oleg-jukovec reviewed (commented):
Please, update the README.md.
We need notes about the `utube`/`utubettl` configuration, policy/option (I'm still not sure about the naming) differences and performance.

Please, apply fixes for the comments to both queues.

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +93,92 @@ function method.normalize_task(self, task)
-    -- See https://www.tarantool.io/ru/doc/latest/concepts/atomic/txn_mode_mvcc/
-
-    if box.cfg.memtx_use_mvcc_engine and (not box.is_in_txn()) then
+    local commit_requirements = box.cfg.memtx_use_mvcc_engine and
+            (not box.is_in_txn())
+    if commit_requirements then
```
If `box.cfg.memtx_use_mvcc_engine == false`, transactions will return an error in case of a yield inside a transaction. That means that, if we have (for example) two parallel `space.index:min` methods inside a transaction block, we will get an `ER_TRANSACTION_CONFLICT` error (in case of `box.cfg.memtx_use_mvcc_engine == false`).

oleg-jukovec replied:
`space.index:min` should not yield in case of the `memtx` engine.

DerekBum replied:
I was trying to do this change, and yes, with `vinyl` I am receiving ` ER_TRANSACTION_CONFLICT: Transaction has been aborted by conflict` error.

oleg-jukovec replied:
> I was trying to do this change, and yes, with `vinyl` I am receiving ` ER_TRANSACTION_CONFLICT: Transaction has been aborted by conflict` error.

It may happen because the `_ready` space has been not created with `vinyl` engine too:
https://github.com/tarantool/queue/blob/80f33dcd352a136c47b9dee9fdb1a4d7bf6e29c0/queue/abstract/driver/utube.lua#L68

To be honest, I don’t understand what options are used here as `space_opts`.

oleg-jukovec replied:
In theory `vinyl` allows `yield`s inside a transaction. Try:

```Lua
box.cfg{}

box.schema.space.create('t2', {engine='vinyl'})
--box.schema.space.create('t2', {engine='memtx'})
box.space.t2:create_index('pk')

box.begin()
box.space.t2:insert({1})
require('fiber').yield()
box.space.t2.index.pk:min()
require('fiber').yield()
box.space.t2:insert({2})
require('fiber').yield()
box.commit()
```

So the problem may be caused by mixing `memtx` + `vinyl` spaces in a single transaction which is not allowed.

DerekBum replied:
Yes, this code works fine.

oleg-jukovec replied:
For now, I suggest to implement the feature only for `memtx` and don't allow to use it with `vinyl`.

Please, create an issue to support the `vinyl` too.

oleg-jukovec replied:
In the case of vinyl and conflict, we will probably just have to try to repeat the transaction.

oleg-jukovec commented:
@DifferentialOrange @better0fdead this is a non-trivial fix, we need additional review with a clear head here.

DerekBum commented on README.md:
```diff
@@ -215,6 +220,11 @@ on disk
   * `on_task_change` - function name - a callback to be executed on every
 operation
+  * `storage_mode` - string - one of
+      * `"default"` - default implementation of `utube`
```
Do we need a common class for storage modes? Or each driver gets its own set of modes? Right now modes for `utube` are inside a `utube.lua` and modes for `utubettl` are inside of `utubettl.lua`. So user cannot access modes as `STORAGE_MODE_DEFAULT` or `STORAGE_MODE_UTUBE_READY_SPACE`.

Or I can move them to the separate file `queue/abstract/driver/modes.lua` and this way user could use them as 
```lua
local modes = require('queue.abstract.driver.modes')

modes.STORAGE_MODE_DEFAULT
```

oleg-jukovec replied:
From my point of view, the implementation of drivers is something from the internal API. So we need the enumeration on the global-level.

oleg-jukovec replied:
https://github.com/tarantool/queue/blob/aa7c092a1f2ac80d3cfc63c75595d10c5a95c908/queue/init.lua#L46

As an alternative we could use it in that manner:
```
queue.driver.utube.STORAGE_MODE_BLA_BLA_BLA
```

It looks a bit better than global variables.

DerekBum commented:
Updated the code according to comments.
Added links to the new issue https://github.com/tarantool/queue/issues/230 about implementing new storage modes for `vinyl` engine.

DerekBum commented on README.md:
```diff
@@ -165,13 +165,49 @@ The main idea of this queue backend is the same as in a `fifo` queue:
 However, tasks may be grouped into sub-queues.
 
+It is advised not to use `utube` methods inside transactions with
+`read-confirmed` isolation level. It can lead to errors when trying to make
```
We use `read-committed` level in the transaction. So if we already in a `read-committed` transaction, everything is fine.
But if we are currently inside a `read-confirmed` transaction, then we get a possible race.

oleg-jukovec replied:
Oh, yes, it is my fault. Sorry.

DerekBum commented on t/benchmark/busy_utubes.lua:
```diff
@@ -0,0 +1,94 @@
+
+            wait_count = wait_count + 1
+        end)
+    end
```
Updated test and benchmark results.

DifferentialOrange commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
     return task
 end
 
+local function take_ready(self)
```
Is it that had to implement a single take function?

DifferentialOrange commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
 -- take task
 function method.take(self)
+    if self.ready_space then
+        return take_ready(self)
```
Let's extract default take to a separate function as well.

DifferentialOrange commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
+                self.space_ready:delete(id)
+
+                if commit_requirements then
+                    box.commit()
```
Three `box.commit`s smells a bit bad. I think you can rewrite this function a bit.

DifferentialOrange commented on queue/abstract/driver/utube.lua:
```diff
@@ -141,49 +282,152 @@ function method.touch(self, id, ttr)
+        if self.ready_space then
+            -- We guarantee that release is called only on TAKEN tasks.
+            local inserted, err =
+                pcall(self.space_ready.insert, self.space_ready, {id, task[3]})
```
In which cases we are expected to fail here?

DerekBum replied:
Yes, we guarantee that release is called only on TAKEN tasks, so this task_id and this utube could not be in the space_ready. This means that there could be no errors in a standard work flow. But, for example, user can call `release` method on an already `READY` task. So we need a protection in case of incorrect usage of the tube.

DifferentialOrange replied:
> user can call release method on an already `READY` task
So it means you don't actually "guarantee that release is called only on `TAKEN` tasks", you expect it to be so. And isn't it easier to just call "replace" in this case?

oleg-jukovec replied:
`space_ready:replace` will break the logic. We need to try to insert and fail on the error. Another ability here:

But it seems we need to update the `_ready` space on the error case + if task_id < task_id in the ready space.

DifferentialOrange replied:
Well, ok. Let's update the comment at least

DifferentialOrange commented on queue/abstract.lua:
```diff
@@ -297,7 +297,11 @@ function tube.drop(self)
+    if self.raw.drop ~= nil then
+        self.raw:drop()
+    else
+        box.space[space_name]:drop()
```
```suggestion
        local space_name = tube[3]
        box.space[space_name]:drop()
```
and remove the one above.

DifferentialOrange reviewed (approved):
The idea seems rather clear after inspecting the issue and, since tests are green, everything should be fine. I had left some code nits, but it shouldn't be anything crucial there.

oleg-jukovec commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
+        transaction_opts = {txn_isolation = 'read-committed'}
+    end
+
+    local commit_requirements = not box.is_in_txn()
```
We need to make multiple updates with box.atomic/box.begin+box.commit to avoid `yield`s between operations (on WAL write).

oleg-jukovec replied:
In the default mode too:
https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_txn_management/begin/

"Disable [implicit yields](https://www.tarantool.io/en/doc/latest/concepts/coop_multitasking/#app-implicit-yields) until the transaction ends."

oleg-jukovec replied:
As example, without a transaction in the default mode:

```
update
commit - WAL write - yield
insert
commit - WAL write - yield
```
With a transaction:
```
begin
update
insert
commit - WAL write - yield
```

oleg-jukovec replied:
We don't need to wrap a single `read` + a single `write` in a transaction with `memtx` + default mode (if we have such places in the code).
But this should not produce an overhead, so it is easier to use one approach everywhere.

DifferentialOrange replied:
> with memtx + default mode
Yeah, I was talking about it. Having transactions everywhere also seems ok.

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +118,112 @@ function method.normalize_task(self, task)
 function method.put(self, data, opts)
-    local max
-
-    -- Taking the maximum of the index is an implicit transactions, so it is
```
It is still relevant, but now we cover entire `put` method in a transaction, not only `min` call.

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,37 +131,120 @@ function method.normalize_task(self, task)
-    else
-        max = self.space.index.task_id:max()
+    local transaction_opts = {}
+    if box.cfg.memtx_use_mvcc_engine then
```
No, unfortunately it's not possible to check mvcc with `box.info`.

DifferentialOrange replied:
Ok, fine

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -141,49 +264,160 @@ function method.touch(self, id, ttr)
+        if self.ready_space then
+            -- We guarantee that release is called only on TAKEN tasks.
+            local inserted, err =
+                pcall(self.space_ready.insert, self.space_ready, {id, task[3]})
```
There is no comments in the Lua style guide (https://www.tarantool.io/en/doc/latest/dev_guide/lua_style_guide/).
So all line breaks are made to my liking.

DerekBum commented on queue/abstract/driver/utube.lua:
```diff
@@ -57,13 +72,56 @@ function tube.create_space(space_name, opts)
         space          = space,
+        space_ready    = space_ready,
         on_task_change = on_task_change,
+        ready_space    = ready_space,
```
Seems more compact with already declared flag:
```lua
if self.ready_space then
    ...
end
```
instead of
```lua
if self.opts.storage_mode == tube.STORAGE_MODE_UTUBE_READY_SPACE then
    ...
end
```

DifferentialOrange commented on queue/abstract/driver/utube.lua:
```diff
@@ -57,13 +74,41 @@ function tube.create_space(space_name, opts)
+    -- TODO <insert issue here>.
+    local ready_space = (opts.storage_mode == storage_modes.MODE_UTUBE_READY_SPACE or false)
+            and opts.engine ~= 'vinyl'
+    if ready_space then
```
Yeah, it sounds much clear

oleg-jukovec replied:
It looks like a good idea to rename the `<tube_name>_ready` space to `<tube_name>_ready_buffer`.

oleg-jukovec replied:
And storage mode from `MODE_UTUBE_READY_SPACE` to `STORAGE_MODE_READY_BUFFER`.

DerekBum commented:
Updated the code according to the comments.

oleg-jukovec reviewed (changes requested):
In the `utubettl` we don't take into account priority when working with STORAGE_MODE_READY_BUFFER.

oleg-jukovec commented on queue/abstract/driver/utubettl.lua:
```diff
@@ -87,18 +102,83 @@ function tube.create_space(space_name, opts)
+            return
+        end
+        if current_task ~= nil and current_task[i_pri] > pri then
+            self.space_ready_buffer:delete(current_task[id])
```
Do the delete not throw an error if tuple does not exists? If it just do nothing - it's ok.

DerekBum replied:
No, there will be no error: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_space/delete/ (according to the example nil will be returned).

oleg-jukovec commented on queue/abstract/driver/utubettl.lua:
```diff
@@ -87,18 +102,83 @@ function tube.create_space(space_name, opts)
+        if current_task == nil or current_task[i_status] ~= state.READY or current_task[i_pri] < pri then
+            return
+        end
+        if current_task ~= nil and current_task[i_pri] > pri then
```
`>` or `<`? Please, ensure that priority works as before.
```suggestion
        if current_task ~= nil and current_task[i_pri] > pri then
```

DerekBum replied:
Priority works the same as nice number -- less numbers have more priority. So if in the ready space currently exists a task with higher number (and thus, lower priority), we need to delete it before putting new task (with higher priority).

oleg-jukovec commented on queue/abstract/driver/utubettl.lua:
```diff
@@ -190,22 +298,70 @@ end
+            space_ready_buffer:create_index('pri', {
+                type = 'tree',
+                parts = {3, num_type(), 1, num_type()},
+                unique = true,
```
It's a side-effect and have not required by our code.
```suggestion
```

oleg-jukovec commented on queue/abstract/driver/utube.lua:
```diff
@@ -193,6 +449,19 @@ function method.kick(self, count)
         task = self.space:update(task[1], {{ '=', 2, state.READY }})
+        if self.ready_space_mode then
+            local prev_task = self.space_ready_buffer.index.utube:get{task[3]}
+            if prev_task ~= nil then
+                delete_ready(self, prev_task[1], task[3])
+            else
```
```suggestion
            if prev_task ~= nil then
                if prev_task[1] > task[1] then -- if id of the exist task > kicked task.
                     delete_ready(self, prev_task[1], task[3])
                end
            else
```

DerekBum replied:
Also did similar change for `utubettl` (but comparing pairs `{pri, id}`).

DerekBum commented:
Updated the code according to comments.

oleg-jukovec commented on queue/abstract/driver/utubettl.lua:
```diff
@@ -87,18 +102,99 @@ function tube.create_space(space_name, opts)
+    local prev_task = self.space_ready_buffer.index.utube:get{utube}
+    if prev_task ~= nil then
+        if prev_task[3] > pri or (prev_task[3] == pri and prev_task[1] > id) then
+            delete_ready(self, prev_task[1], utube)
```
It seems too heavy for the update:
```suggestion
            self.space_ready_buffer:delete(id)
            self.space_ready_buffer:insert({id, utube, pri})
```

oleg-jukovec commented on queue/abstract/driver/utube.lua:
```diff
@@ -73,43 +131,128 @@ function method.normalize_task(self, task)
 
--- take task
-function method.take(self)
+-- take the first task form the ready_buffer.
```
```suggestion
-- Take the first task form the ready_buffer.
```

oleg-jukovec commented on queue/abstract/driver/utubettl.lua:
```diff
@@ -293,14 +466,60 @@ function method.touch(self, id, delta)
 
--- take task
-function method.take(self)
+-- take the first task form the ready_buffer.
```
```suggestion
-- Take the first task form the ready_buffer.
```
